openedx/openedx-platform
# Add OpenStack Swift support

**Background:** The edX platform includes a number of features that use S3 directly. These include:
- Instructor CSV data exports
- Video storage
- xqueue
- Student Identity Verification (photo uploads)
- Analytics event dumps (via a logrotate script)

For deployments outside of AWS it would make sense to use a different cloud storage provider, or even the local filesystem.

In particular, OpenStack deployments will want to use an OpenStack Swift Object Store for user-uploaded content. Although Swift offers S3 API compatibility as a plugin (called "Swift3") in theory, in practice the compatibility is limited, often not installed on the OpenStack instance, and hard to get working.

This pull request is an attempt to solve this problem by replacing all S3-specific code with the [django storage API](https://docs.djangoproject.com/en/1.8/ref/files/storage/).

**Studio updates:** None yet, although video uploads will need to be modified. I am currently holding off on this as it will require changes to VEDA and I don't have access to that repo. Once Greg Martin has made the necessary changes to VEDA I will update Studio in a separate PR.

**LMS updates:** Instructor CSV exports and student photo uploads modified to use the django storage API. Added an `openstack.py` settings file.

**Sandbox:** http://pr11286.sandbox.opencraft.com/

**Related PRs:**
- https://github.com/edx/configuration/pull/2723
- https://github.com/edx/xqueue/pull/103
- https://github.com/edx/edx-platform/pull/11374

**Testing:** Deploy a sandbox using https://github.com/edx/configuration/pull/2723. Use [these ansible vars](https://gist.github.com/82623f60243d3ae94152), completed with your openstack credentials.

**To test the instructor CSV reports:**
- Go to the instructor dashboard for a course
- Open the 'data download' tab
- Click 'generate grade report'
- An entry for the report should appear in the list below
- Check that the report CSV has been uploaded to the swift container
- Click the link to download the report directly from swift

**To test student identity verification:**
- In the django admin, create a new `CourseMode` for an existing course. Set the mode to `verified`, and make sure the price is > 0 (https://github.com/edx/edx-platform/pull/11374)
- In the LMS course list, there should now be an 'upgrade' button if you are already enrolled in the course. Click it
- You will be presented with a payment page. We need to get past this to get to the ID verification step. Run `./manage.py lms --settings=openstack shell`, then:

``` python
from django.contrib.auth.models import User
from opaque_keys.edx.keys import CourseKey
from student.models import CourseEnrollment

user = User.objects.get(username='<your_username>')
course_key = CourseKey.from_string('course-v1:edX+DemoX+Demo_Course')
enrollment = CourseEnrollment.get_or_create_enrollment(user, course_key)
enrollment.update_enrollment(mode='verified')
```

Alternatively, you can simply update the `CourseEnrollment` in the django admin:

<img width="366" alt="screen shot 2016-02-01 at 18 44 59" src="https://cloud.githubusercontent.com/assets/793379/12718009/3cc7d1e2-c91e-11e5-8bb0-ce1f1fdffabb.png">

This fools the LMS into thinking that the course has already been paid for. Go back to the payment page, and you should see the webcam verification step. Note that Chrome will not allow access to the webcam over plain http unless you are running on localhost. Firefox doesn't seem to mind.

**To test xqueue:**
- In studio, add a blank problem and set its xml to:

``` xml
<problem display_name="XQueue Test">
   <text>
       <p>Upload a program that prints "hello world".</p>
   </text>
   <coderesponse queuename="edX-Open_DemoX">
       <filesubmission/>
       <codeparam>
           <initial_display>
             # students please write your program here
             print ""
           </initial_display>
           <answer_display>
             print "hello world"
           </answer_display>
           <grader_payload>
             {"output": "hello world"}
           </grader_payload>
       </codeparam>
   </coderesponse>
</problem>
```
- Go to that problem **in LMS** (xqueue does not work in studio), select a file and click the 'Check' button
- You should see this:
  <img width="812" alt="screen shot 2016-02-03 at 01 41 57" src="https://cloud.githubusercontent.com/assets/793379/12760245/09dec876-ca18-11e5-8ec8-85f57a93669a.png">
- The submission should now be in the configured swift bucket

**To test log sync to swift:**

SSH into the sandbox, and run

```
sudo service supervisor stop
```

This should trigger a log sync, after which the tracker logs can be found in a swift container.

**Future tasks:** write some tests for django-swift-storage. It doesn't have any yet.

**Settings**

``` yaml
EDXAPP_SETTINGS: 'openstack'
XQUEUE_SETTINGS: 'openstack_settings'
COMMON_VHOST_ROLE_NAME: 'openstack'
edx_ansible_source_repo: 'https://github.com/open-craft/configuration.git'
configuration_version: 'omar/openstack'
xqueue_source_repo: 'https://github.com/open-craft/xqueue.git'
xqueue_version: 'omar/django-storage-api'
```
